File: src/registry.ts
import fs from "node:fs";
import path from "node:path";

import type { AppConfig, Project, ProjectType, ScriptDefinition } from "./models.ts";
import { parseSimpleToml } from "./toml.ts";

const MANIFEST_NAME = "manifest.toml";

interface RawManifest {
  id?: string;
  name?: string;
  description?: string;
  project_types?: ProjectType[];
  entry?: string;
  default_args?: Record<string, unknown>;
}

export function loadScripts(config: AppConfig): ScriptDefinition[] {
  const scriptsDir = config.scripts.directory;
  if (!fs.existsSync(scriptsDir)) {
    return [];
  }
  const entries = fs.readdirSync(scriptsDir, { withFileTypes: true }).sort((a, b) => a.name.localeCompare(b.name));
  const scripts: ScriptDefinition[] = [];
  for (const entry of entries) {
    if (!entry.isDirectory()) {
      continue;
    }
    const directory = path.join(scriptsDir, entry.name);
    const manifestPath = path.join(directory, MANIFEST_NAME);
    if (!fs.existsSync(manifestPath)) {
      continue;
    }
    scripts.push(loadManifest(directory, manifestPath));
  }
  return scripts;
}

export function applicableScripts(scripts: ScriptDefinition[], projects: Project[]): ScriptDefinition[] {
  if (projects.length === 0) {
    return scripts;
  }
  return scripts.filter((script) =>
    projects.every((project) => script.projectTypes.some((value) => project.projectTypes.includes(value))),
  );
}

export function getScriptById(scripts: ScriptDefinition[], scriptId: string): ScriptDefinition | undefined {
  return scripts.find((script) => script.scriptId === scriptId);
}

function loadManifest(directory: string, manifestPath: string): ScriptDefinition {
  const raw = parseSimpleToml(fs.readFileSync(manifestPath, "utf8")) as RawManifest;
  const missing = ["id", "name", "description", "project_types", "entry"].filter((key) => !(key in raw));
  if (missing.length > 0) {
    throw new Error(`Missing fields in ${manifestPath}: ${missing.join(", ")}`);
  }
  return {
    scriptId: String(raw.id),
    name: String(raw.name),
    description: String(raw.description),
    projectTypes: [...(raw.project_types ?? [])],
    entry: String(raw.entry),
    directory,
    manifestPath,
    defaultArgs: { ...(raw.default_args ?? {}) },
  };
}

File: src/devtools/registry.py
from __future__ import annotations

import tomllib
from pathlib import Path

from devtools.models import AppConfig, Project, ScriptDefinition


MANIFEST_NAME = "manifest.toml"


def load_scripts(config: AppConfig) -> list[ScriptDefinition]:
    scripts_dir = config.scripts.directory
    if not scripts_dir.exists():
        return []

    scripts: list[ScriptDefinition] = []
    for child in sorted(scripts_dir.iterdir()):
        if not child.is_dir():
            continue
        manifest_path = child / MANIFEST_NAME
        if not manifest_path.exists():
            continue
        scripts.append(_load_manifest(child, manifest_path))
    return scripts


def applicable_scripts(scripts: list[ScriptDefinition], projects: list[Project]) -> list[ScriptDefinition]:
    if not projects:
        return scripts
    return [
        script
        for script in scripts
        if all(set(script.project_types).intersection(set(project.project_types or [project.project_type])) for project in projects)
    ]


def get_script_by_id(scripts: list[ScriptDefinition], script_id: str) -> ScriptDefinition | None:
    for script in scripts:
        if script.script_id == script_id:
            return script
    return None


def _load_manifest(directory: Path, manifest_path: Path) -> ScriptDefinition:
    with manifest_path.open("rb") as handle:
        raw = tomllib.load(handle)

    required = ("id", "name", "description", "project_types", "entry")
    missing = [key for key in required if key not in raw]
    if missing:
        missing_fields = ", ".join(missing)
        raise ValueError(f"Missing fields in {manifest_path}: {missing_fields}")

    return ScriptDefinition(
        script_id=str(raw["id"]),
        name=str(raw["name"]),
        description=str(raw["description"]),
        project_types=[str(item) for item in raw["project_types"]],
        entry=str(raw["entry"]),
        directory=directory,
        manifest_path=manifest_path,
        default_args=dict(raw.get("default_args", {})),
    )
